lucene/core/src/java/org/apache/lucene/index/IndexSorter.java

```diff
@@ -0,0 +1,500 @@
+
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.store.DataOutput;
+import org.apache.lucene.util.LongValues;
+import org.apache.lucene.util.NumericUtils;
+import org.apache.lucene.util.packed.PackedInts;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * Handles how documents should be sorted in an index, both within a segment and between
+ * segments.
+ *
+ * Implementers must provide the following methods:
+ * {@link #getDocComparator(LeafReader)} - an object that determines how documents within a segment are to be sorted
+ * {@link #getComparableProviders(List)} - an array of objects that return a sortable long value per document and segment
+ * {@link #serialize(DataOutput)} - how the sort should be written into the segment header
+ * {@link #getProviderName()} - the SPI-registered name of a {@link SortFieldProvider} to deserialize the sort
+ *
+ * The companion {@link SortFieldProvider} should be registered with SPI via {@code META-INF/services}
+ */
+public interface IndexSorter {
+
+  /** Used for sorting documents across segments */
+  public interface ComparableProvider {
+    /**
+     * Returns a long so that the natural ordering of long values matches the
+     * ordering of doc IDs for the given comparator
+     */
+    long getAsComparableLong(int docID) throws IOException;
+  }
+
+  /** A comparator of doc IDs, used for sorting documents within a segment */
+  public interface DocComparator {
+    /** Compare docID1 against docID2. The contract for the return value is the
+     *  same as {@link Comparator#compare(Object, Object)}. */
+    int compare(int docID1, int docID2);
+  }
+
+  /**
+   * Get an array of {@link ComparableProvider}, one per segment, for merge sorting documents in different segments
+   * @param readers the readers to be merged
+   */
+  public abstract ComparableProvider[] getComparableProviders(List<? extends LeafReader> readers) throws IOException;
+
+  /**
+   * Get a comparator that determines the sort order of docs within a single Reader.
+   *
+   * NB We cannot simply use the {@link FieldComparator} API because it requires docIDs to be sent
+   * in-order. The default implementations allocate array[maxDoc] to hold native values for comparison,
+   * but 1) they are transient (only alive while sorting this one segment) and 2) in the typical
+   * index sorting case, they are only used to sort newly flushed segments, which will be smaller
+   * than merged segments
+   *
+   * @param reader the Reader to sort
+   */
+  public abstract DocComparator getDocComparator(LeafReader reader) throws IOException;
+
+  /**
+   * Serializes the parent SortField.  This is used to write Sort information into the Segment header
+   *
+   * @see SortFieldProvider#loadSortField(DataInput)
+   */
+  public abstract void serialize(DataOutput out) throws IOException;
```

Comment:
I wonder if it would work to move it to SortFieldProvider instead for symmetry, so that the serialization and deserialization logic would be in the same place.